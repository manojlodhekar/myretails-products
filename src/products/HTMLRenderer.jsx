import React from 'react';

const HTMLRenderer = (props) =>{
    return(
        <div>
            <div style={{paddingTop: 10}}>
                <span id="htmlRendererId" className="returnPolicyCls" dangerouslySetInnerHTML={{__html: props.htmlContents}} />
            </div>
        </div>
    )
}
export default HTMLRenderer;