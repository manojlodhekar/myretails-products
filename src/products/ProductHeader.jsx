import React from 'react';
import { Image, } from 'semantic-ui-react'
import productConst from '../const/productConst';

const styles = {
  
    header:{
        paddingLeft: 10,
        textAlign: 'left',
        paddingBottom: 10,
        borderBottom: '1px solid grey'
    },
    headerTxt:{
        paddingLeft: 10,
        textAlign: 'left',
        fontSize: 24,
        fontWeight: 'bold',
        lineHeight: 1.2,
        color: 'white'
    },
    logoCls:{
        maxWidth: 55, 
        float: 'right', 
        textAlign: 'right'
    },
    subHeaderText:{
        paddingLeft: 10,
        textAlign: 'left',
        fontSize: 18,
        color: 'white'
    },
    headMessage:{
        width: '100%'
    },
    cartIcon:{
        backgroundColor: '#fff',
       //margin: '2px 8px 2px 5px'
    },
    cart:{
        
    }

}


const ProductHeader = (props) =>{

    return(
        <div className="headContainer" style={{display: 'flex'}}>
            <div>
                <Image src={productConst.TARGET_LOGO} style={styles.logoCls} />
            </div>
            <div style={styles.headMessage}>
                <div style={styles.headerTxt}>
                        {productConst.HEADER_TEXT}
                </div>
                <div style={styles.subHeaderText}>
                        {props.subHeaderText}
                </div>
            </div>
        </div>
    )
}
export default ProductHeader;