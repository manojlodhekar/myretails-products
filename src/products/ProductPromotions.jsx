import React from 'react';
import { Icon, Divider } from 'semantic-ui-react';

const ProductPromotions = (props) =>{

    const renderPromotionDetails = () =>{

        let productDetail =  props.promotionDetails.map((item) => {
            return(
                <div className="promotions" key={item.promotionIdentifier}>
                    <div>
                        <Icon name='tag'/>{item.Description[0].shortDescription} 
                    </div>
                </div>
            )
        });
        return productDetail;
    }

    return(
        <div>
            <Divider />
            {renderPromotionDetails()}
            <Divider />
        </div>
    )
}
export default ProductPromotions;