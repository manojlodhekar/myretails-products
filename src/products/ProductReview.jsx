import React, { Component } from 'react';
import { Grid, Divider } from 'semantic-ui-react';
import CustomerReview from './CustomerReview';
import { getStarRating} from '../utils/ProductUtils';
import productConst from '../const/productConst';


class ProductReview extends Component {

    constructor(props){
        super(props);
        this.state = {
            showReviewFlag: false
        }
    }
    showHideReview = () =>{
        this.setState({
            showReviewFlag : !this.state.showReviewFlag
        })
    }
    render(){
        
        const {ProductReviewDetail} = this.props;
        return(
            <div >
                <div>
                    <span className="ratingTxtCls"> 
                        <span className="starCls"> {getStarRating(5)} </span> 
                        <span>Overall</span>  
                        <span className="viewReviewCls"
                        onClick={(e) =>this.showHideReview()}> View all {ProductReviewDetail.Reviews.length} Reviews </span> 
                    </span>
                </div>
                <div>
                    <Grid divided='vertically' className="reviewContainer">
                        
                        <Grid.Row columns={2}>
                        <Grid.Column>
                            <div>
                                <div className="reviewTitleCls">PRO</div>
                                <div className="reviewTextCls">{productConst.PRO_HELPFUL}</div>
                                <Divider style={{width:'125%'}}/>
                                <CustomerReview customerReview={ProductReviewDetail.Pro[0]} />
                            </div>
                        </Grid.Column>
                        <Grid.Column>
                            <div>
                            <div className="reviewTitleCls">CON</div>
                                <div className="reviewTextCls">{productConst.CON_HELPFUL}</div>
                                <Divider style={{width:'110%'}}/>
                            <CustomerReview customerReview={ProductReviewDetail.Con[0]} />
                            </div>
                        </Grid.Column>
                        </Grid.Row>
                    </Grid>
                        {
                            this.state.showReviewFlag ?
                            <Grid divided='vertically' className="reviewContainer">
                            <Grid.Row columns={1}>
                            <Grid.Column>
                                <div>
                                    <div className="reviewTitleCls">
                                        All other Reviews
                                        <span className="viewReviewCls"
                                         onClick={(e) => this.showHideReview()}> Hide Review </span>
                                    </div>
                                    <Divider style={{width:'100%'}}/>
                                    {
                                        ProductReviewDetail.Reviews.map((review,index) =>{
                                            return (
                                            <div key={index}>
                                                <CustomerReview customerReview={review} />
                                                <Divider style={{width:'100%'}}/>
                                            </div>
                                            )
                                        })
                                    }
                                    <span className="viewReviewCls"
                                         onClick={(e) => this.showHideReview()}> Hide Review </span>
                                </div>
                            </Grid.Column>
                            </Grid.Row> 
                        </Grid>
                            : null
                        }
                        
                </div>
            </div>
        )
    };
}
export default ProductReview;