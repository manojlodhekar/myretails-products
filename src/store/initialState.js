export default {
    products: {
        selectedProductDetail: {},
        productDetailImages:{},
        productDetailDescription:{},
        productDetailCustomerReview:{},
        productDetailOffers:{},
        productDetailPromotions:[],
        productDetailReturnPolicy:{},
        productApiError: false
    }
}