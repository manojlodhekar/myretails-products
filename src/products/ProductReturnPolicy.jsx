import React from 'react';
import HTMLRenderer from './HTMLRenderer';

const ProductReturnPolicy = (props) =>{

    const renderProductReturnPolicy = () =>{

        let returnPolicy =  props.returnPolicyDetails.ReturnPolicyDetails.map((policy, index) => {
            return(
                <div className="returnPolicyCls" key={index}> 
                    { 
                        index === 0 ? <div className="policyHeader">{policy.guestMessage} : </div>  : null
                    }
                    
                    <div>
                    {policy.user} : {policy.policyDays} Days
                    </div>
                </div>
            )
        });
        return returnPolicy;
    }

    return(
        <div>
            {renderProductReturnPolicy()}
            <div style={{paddingTop: 10}}>
                <HTMLRenderer htmlContents={props.returnPolicyDetails.legalCopy} />
            </div>
        </div>
    )
}
export default ProductReturnPolicy;