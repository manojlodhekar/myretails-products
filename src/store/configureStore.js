import { createStore, applyMiddleware } from 'redux';
import rootReducer from '../reducers';
import thunk from 'redux-thunk';    
import initialState from './initialState'

export default function configureStore() {
    return (createStore(
        rootReducer,initialState,applyMiddleware(thunk)
    ));
}