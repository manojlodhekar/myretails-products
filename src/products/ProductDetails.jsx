import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Image, Item, Icon, Grid, Segment} from 'semantic-ui-react';
import PropTypes from 'prop-types';
import { getProductDetails } from '../actions/getProductDetails';
import ProductHeader from './ProductHeader';
import productConst from '../const/productConst';
import ProductPromotions from './ProductPromotions';
import ProductReturnPolicy from './ProductReturnPolicy'
import HTMLRenderer from './HTMLRenderer';
import ProductReview from './ProductReview';
import ProductCarousel from './ProductCarousel';
import ProductsZoomDialog from './ProductsZoomDialog';

class ProductDetails extends Component {
    
    constructor(props){
        super(props);
        this.state = {
            quantity : 1,
            windowWidth:window.innerWidth,
            currentActiveIndex: 0,
            open: false
        }
        this.updateDimensions = this.updateDimensions.bind(this);
        this.buyProduct = this.buyProduct.bind(this);
    };

    componentDidMount(){
        this.props.actions.getProductDetails();
        window.addEventListener("resize", this.updateDimensions);
    }

    buyProduct = (btnStatus) =>{
        alert(` ${this.state.quantity} items added to cart.`);
    }
    updateDimensions() {
        const realContainerWidth = window.innerWidth;
        this.setState({
            windowWidth:realContainerWidth 
        })
      }
    changeQuantity = (operation) => {
        if(operation === 'remove'){
            if(this.state.quantity === 1)
            return;
            this.setState({ quantity: this.state.quantity - 1 })
        }
        else{
            this.setState({ quantity: this.state.quantity + 1})
        }
      }

      setHtmlContent = (content) =>{
          if(!content)return;

          let productDesc = content.map(item => {
            return(`<li>${item}</li>`)
          });

          return productDesc;
      }

      selectRetailsProductImage = (index , url) =>{
        console.log('index = ' + index + " URL" + url);
        this.setState({currentActiveIndex: index})
      }
      openZoomedImage = () =>{
        this.setState({ open: true })
      }
      modalPopupClosed = () =>{
        this.setState({
          open: false
        })
      }
      getCurrentImageUrl = () =>{
          if(this.props.productDetailImages.AlternateImages){
              return this.props.productDetailImages.AlternateImages;
            }
          
      }

    render(){

        const {selectedProductDetail, productDetailImages, productDetailDescription, productDetailCustomerReview, productDetailOffers,
            productDetailPromotions, productDetailReturnPolicy} = this.props;
            
        let productsDesc =  this.setHtmlContent(productDetailDescription.features);
        productsDesc = productsDesc ? `<ul> ${productsDesc.join('')} </ul>` : '';

        if(selectedProductDetail.CatalogEntryView && productDetailImages.PrimaryImage  && !this.imageCounter){
            this.imageCounter = !this.imageCounter;
            productDetailImages.AlternateImages.unshift(productDetailImages.PrimaryImage[0]);
        }
        
        return(
            
                <div className="detailsContainer">
                    <ProductHeader subHeaderText={productConst.PRODUCT_MAIN_PAGE_SUBHEADER} />
                {
                    selectedProductDetail.itemId ?
                    <div>
                        <Item.Group>
                            <Item>
                            <div className="productLeft">
   
                                <div className="productTitle">{selectedProductDetail.title}</div> 
                                <Item.Image size='large' src={productDetailImages.AlternateImages[this.state.currentActiveIndex].image}
                                style={{paddingLeft: '10%'}}/>
                                <div className="viewLargeIcon" onClick={(e) => this.openZoomedImage()}>
                                    <Icon name='zoom' /> View Larger
                                </div>
                                <ProductCarousel productsList={productDetailImages.AlternateImages} currentActiveIndex={this.state.currentActiveIndex} 
                                selectProductImage = {this.selectRetailsProductImage} />
                                {
                                    this.state.windowWidth >= 768 ?
                                    <ProductReview ProductReviewDetail={productDetailCustomerReview} />:
                                    null
                                }
                            
                            </div>
                        <Item.Content className='productContent'>
                        <Item.Meta className='productPrice'>{productDetailOffers.OfferPrice[0].formattedPriceValue} <span className="offerPriceTxt"> {productDetailOffers.OfferPrice[0].priceQualifier}</span> </Item.Meta>
                            <ProductPromotions promotionDetails={productDetailPromotions} />
                            <div className="quantityCls">
                                <span>Quantity: </span>  <span className="quantityIconCls"> <Icon name="minus circle" onClick={(e) => this.changeQuantity('remove')}/> <span className="category">{this.state.quantity} </span> <Icon name="plus circle" onClick={(e) =>this.changeQuantity('add')}/> </span>
                            </div>
                                {
                                (selectedProductDetail.inventoryStatus === 'Online' && (selectedProductDetail.purchasingChannelCode === '0' || selectedProductDetail.purchasingChannelCode === '2')) ?
                                    <Image src={productConst.ADD_TO_CART} className="buttonCls" onClick={ (e) => this.buyProduct(productConst.CART_BTN_LABEL)} /> 
                                    :
                                    <div>
                                        <Image src={productConst.PICK_IN_STORE} className="buttonCls" onClick={(e) => this.buyProduct(productConst.PICK_BTN_LABEL)} />
                                        <span style={{padding: 50}}>{productConst.FIND_STORE} </span>
                                    </div>
                                }

                            <ProductReturnPolicy returnPolicyDetails={productDetailReturnPolicy} />
                            <div style={{width: '100%', paddingLeft: 25, paddingBottom: 15}}>

                            <Grid columns={3}>
                                <Grid.Row>
                                <Grid.Column className="segCol">
                                     <Segment className="segmentItem" onClick={()=>alert(`Click to action for "${productConst.ADD_TO_REG}"`)}>{productConst.ADD_TO_REG}</Segment>
                                </Grid.Column>
                                <Grid.Column className="segCol">
                                     <Segment className="segmentItem" onClick={()=>alert(`Click to action for "${productConst.ADD_TO_LIST}"`)}>{productConst.ADD_TO_LIST}</Segment>
                                </Grid.Column>
                                <Grid.Column className="segCol">
                                     <Segment className="segmentItem" onClick={()=>alert(`Click to action for "${productConst.SHARE}"`)}>{productConst.SHARE}</Segment>
                                </Grid.Column>
                                </Grid.Row>
                            </Grid>

                            </div>
                            <div>
                                <span className="productTitle">{productConst.PRODUCT_HIGHLIGHT}</span>
                                    <HTMLRenderer htmlContents={productsDesc} />
                            </div>
                            {
                            this.state.windowWidth < 768 ?
                            <ProductReview ProductReviewDetail={productDetailCustomerReview} />:
                            null
                            }

                            
                            </Item.Content>
                        </Item>
                    </Item.Group>
                    <ProductsZoomDialog openPopup={this.state.open}
        imageURLs={this.getCurrentImageUrl()}
        popupClosed = {this.modalPopupClosed}
        currentImgIndex={this.state.currentActiveIndex}/>
                    </div>
                    :
                    <div> Loading....</div>
                }
                
        </div>
        );
    };
}

ProductDetails.propTypes = {
    selectedProductDetail: PropTypes.object.isRequired,
    productDetailImages: PropTypes.object.isRequired,
    productDetailDescription: PropTypes.object.isRequired,
    productDetailCustomerReview: PropTypes.object.isRequired,
    productDetailOffers: PropTypes.object.isRequired,
    productDetailPromotions: PropTypes.array.isRequired,
    productDetailReturnPolicy: PropTypes.object.isRequired,
    actions: PropTypes.object.isRequired
}
function mapStateToProps(state){
    return{
        selectedProductDetail: state.products.selectedProductDetail,
        productDetailImages: state.products.productDetailImages,
        productDetailDescription: state.products.productDetailDescription,
        productDetailCustomerReview: state.products.productDetailCustomerReview,
        productDetailOffers: state.products.productDetailOffers,
        productDetailPromotions: state.products.productDetailPromotions,
        productDetailReturnPolicy: state.products.productDetailReturnPolicy
    };
}

function mapDispatchToProps(dispatch){
    return {
        actions: bindActionCreators( { getProductDetails }, dispatch),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(ProductDetails);
