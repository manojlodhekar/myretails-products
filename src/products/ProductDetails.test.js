import React from 'react';
import configureStore from 'redux-mock-store';
import ProductDetails from './ProductDetails';
import ProductHeader from './ProductHeader';
import ProductPromotions from './ProductPromotions';
import ProductReturnPolicy from './ProductReturnPolicy';
import HTMLRenderer from './HTMLRenderer';
import ProductReview from './ProductReview';

import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount } from 'enzyme';
import initialState from '../store/initialState'

jest.mock('react-dom');

Enzyme.configure({ adapter: new Adapter() });
const mockStore = configureStore();
const store = mockStore(initialState);
const obj = {a: "test1", b: "test2"}
let productDetails = mount(
    <ProductDetails store={store}>
       <ProductHeader />
       <ProductPromotions promotionDetails={obj} />
       <ProductReturnPolicy returnPolicyDetails={obj} />
       <HTMLRenderer htmlContents={obj} />
       <ProductReview ProductReviewDetail={obj} />

    </ProductDetails>
);

describe('ProductDetails is defined', () => {

  it('should be defined', () => {
    expect(productDetails).toBeDefined();
  });
  it('should render correctly', () => {
    expect(productDetails).toHaveLength(1);
  });

  it('should render the productHeader component', () => {
    const productHeader = productDetails.find(ProductHeader);
    expect(productDetails.find(productHeader)).toBeDefined();
  });
  it('should render the ProductPromotions component', () => {
    const productPromotions = productDetails.find(ProductPromotions);
    expect(productDetails.find(productPromotions)).toBeDefined();
  });
  it('should render the ProductReturnPolicy component', () => {
    const productReturnPolicy = productDetails.find(ProductReturnPolicy);
    expect(productDetails.find(productReturnPolicy)).toBeDefined();
  });
  it('should render the HTMLRenderer component', () => {
    const hTMLRenderer = productDetails.find(HTMLRenderer);
    expect(productDetails.find(hTMLRenderer)).toBeDefined();
  });
  it('should render the ProductReview component', () => {
    const productReview = productDetails.find(ProductReview);
    expect(productDetails.find(productReview)).toBeDefined();
  });

  it('should render inner component', () => {
    expect(productDetails).toMatchSnapshot();
  });

 });
