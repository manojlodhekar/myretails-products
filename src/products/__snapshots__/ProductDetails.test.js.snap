// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails is defined should render inner component 1`] = `
ReactWrapper {
  "length": 1,
  Symbol(enzyme.__unrendered__): <Connect(ProductDetails)
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
>
    <ProductHeader />
    <ProductPromotions
        promotionDetails={
            Object {
                "a": "test1",
                "b": "test2",
              }
        }
    />
    <ProductReturnPolicy
        returnPolicyDetails={
            Object {
                "a": "test1",
                "b": "test2",
              }
        }
    />
    <HTMLRenderer
        htmlContents={
            Object {
                "a": "test1",
                "b": "test2",
              }
        }
    />
    <ProductReview
        ProductReviewDetail={
            Object {
                "a": "test1",
                "b": "test2",
              }
        }
    />
</Connect(ProductDetails)>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
